[FIX] Remove extraneous quotations around url encoded vp_token

Ryanmtate opened the pull request:
This PR adds a helper method `clean_quoted_string` to strip extra quotations in JSON encoded `vp_token` payload for url encoding.

There might be a better way to serialize the `VpToken` to ensure an extra quotation is not included, open to suggestions.

Drive by changes:

- Adds an optional `state` property to from the authorization request to the authorization response.
- Adds some serde derive macros to help debug.

Ryanmtate committed:
strip extra quotations in authorization response url encoding, add optional state property to response

Signed-off-by: Ryan Tate <ryan.tate@spruceid.com>

cobward reviewed (changes requested):
I'm not certain we should implement it this way. Older versions of the spec require the vp_token to be a JSON encoded string:
https://github.com/spruceid/openid4vp/blob/2210b2c1368f19f639f810f95b801e253c6e635d/src/core/response/parameters.rs#L38

The latest version of the spec removes the "JSON" from that requirement, which is why there is this mismatch between the implementation here and some implementations we see in the wild.

I'm worried that if we make this change, we will unknowingly break current and future implementations. Could you instead use a flag to decide whether or not to use JSON string encoding?

Ryanmtate commented:
> I'm worried that if we make this change, we will unknowingly break current and future implementations. Could you instead use a flag to decide whether or not to use JSON string encoding?

How would we determine this in practice? We know ahead of time with CA, but for other verifiers, hard to know whether it will be accepted or not.

cobward commented:
> > I'm worried that if we make this change, we will unknowingly break current and future implementations. Could you instead use a flag to decide whether or not to use JSON string encoding?
> 
> How would we determine this in practice? We know ahead of time with CA, but for other verifiers, hard to know whether it will be accepted or not.

We can only know by testing. The default behaviour can be to not include the quotation marks (as that is the future of the spec), but I think we need a way to include them. The quotation marks are in the current version because that is what was required for a previous integration.

Ryanmtate committed:
add should_strip_quotes bool option to auth response

This will remove quotations around the `vp_token`
url-encoded parameter, i.e. %22

Signed-off-by: Ryan Tate <ryan.tate@spruceid.com>
